site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md

```diff
@@ -0,0 +1,437 @@
+
+```patch
+<<<<<<< HEAD
+	th := Setup(t).InitBasic()
+||||||| parent of 9a51c73f6... MM-22619: check for nil plugins environment (#13901)
+	th := Setup(t)
+=======
+	th := SetupWithCustomConfig(t, func(config *model.Config) {
+		*config.PluginSettings.Enable = false
+	})
+>>>>>>> 9a51c73f6... MM-22619: check for nil plugins environment (#13901)
+	defer th.TearDown()
+```
+
+I have a mergetool configured with git to show this same information using a three-way merge inside Vim:
+
+```sh
+git mergetool
+```
+
+The middle pane shows the common ancestor. On its left is what's in the current branch, and on the right is what's incoming. The bottom pane is the current file, showing the conflict markers. In general with merge conflicts, you have to understand the context of the code you are changing in order to make the decision on what version to keep, or whether to blend the changes somehow. In this case, I happen to know that a colleague of mine removed a redundant call to `InitBasic` in the master branch as part of a cleanup effort, but those clean up changes aren't in this release branch. So I need to combine the two changes together and put the call to `InitBasic` on top of the my original code. Then I'll get rid of the other sections and the conflict markers.
+
+With the conflicts resolved, I'm good to finish the cherry-pick!
+
+```sh
+git cherry-pick --continue
+```
+
+I've created a new commit that contains the changes from another branch, and would now be ready to push up this branch for review and CI testing before merging it into the release branch.
+
+The conflict resolution I just performed might have just as easily happened when doing a `git revert`. The process would be the same, and the hard part shouldn't be the tool you are using, but knowing the context of the code you are changing so you can make the right decisions.
+
+### git rebase
+
+My all time favourite git command is `git rebase`! Long before I learned git, I used a version control system called Perforce. It was undoubtedly powerful for its time, but I found it intensely frustrating to organize my code changes for review. Let me describe my typical workflow:
+
+```sh
+git checkout feature-branch
+vim ...
+git commit -m 'wip'
+vim ...
+git commit -m 'wip'
+...
+```
+
+This technique is known as "commit early and often." Every now and then, I'll also push my feature branch upstream (without opening a pull request) just to save my changes in case my computer dies. This is a perfectly fine way to develop, but I submit it makes for very poor pull requests. You can read my thoughts on [Submitting Great PRs](https://developers.mattermost.com/blog/submitting-great-prs/), but the short summary is that the best pull requests consist of a sequence of logically ordered commits that break up the changes into easy to follow chunks.
```

Comment:
```suggestion
This technique is known as "commit early and often". Every now and then, I'll also push my feature branch upstream (without opening a pull request) just to save my changes in case my computer dies. This is a perfectly fine way to develop, but it makes for very poor pull requests. You can read my thoughts on [Submitting Great PRs](https://developers.mattermost.com/blog/submitting-great-prs/), but the summary is that the best pull requests consist of a sequence of logically ordered commits that break up the changes into easy to follow chunks.
```